linera-rpc/src/grpc/server.rs

```diff
@@ -399,75 +400,133 @@ where
+                    && rand::thread_rng().gen::<f32>() < cross_chain_sender_failure_rate
+                {
+                    warn!("Dropped 1 cross-chain message intentionally.");
+                    return;
+                }
+
+                let mut i = 0;
+                while i < cross_chain_max_retries {
+                    linera_base::time::timer::sleep(cross_chain_sender_delay).await;
+                    // Delay increases linearly with the attempt number.
+                    if i > 0 {
+                        _permit = None;
+                        linera_base::time::timer::sleep(cross_chain_retry_delay * i).await;
+                        _permit = Some(semaphore_clone.clone().acquire_owned().await.unwrap());
                     }
 
-                    for i in 0..cross_chain_max_retries {
-                        // Delay increases linearly with the attempt number.
-                        linera_base::time::timer::sleep(
-                            cross_chain_sender_delay + cross_chain_retry_delay * i,
-                        )
-                        .await;
-
-                        let result = || async {
-                            let cross_chain_request = cross_chain_request.clone().try_into()?;
-                            let request = Request::new(cross_chain_request);
-                            let mut client =
-                                ValidatorWorkerClient::new(pool.channel(remote_address.clone())?)
-                                    .max_encoding_message_size(GRPC_MAX_MESSAGE_SIZE)
-                                    .max_decoding_message_size(GRPC_MAX_MESSAGE_SIZE);
-                            let response = client.handle_cross_chain_request(request).await?;
-                            Ok::<_, anyhow::Error>(response)
-                        };
-                        match result().await {
-                            Err(error) => {
-                                warn!(
-                                    nickname,
-                                    %error,
-                                    i,
-                                    from_shard = this_shard,
-                                    to_shard = shard_id,
-                                    "Failed to send cross-chain query",
-                                );
+                    let result = || async {
+                        let cross_chain_request = cross_chain_request.clone().try_into()?;
+                        let request = Request::new(cross_chain_request);
+                        let mut client =
+                            ValidatorWorkerClient::new(pool.channel(remote_address.clone())?)
+                                .max_encoding_message_size(GRPC_MAX_MESSAGE_SIZE)
+                                .max_decoding_message_size(GRPC_MAX_MESSAGE_SIZE);
+                        let response = client.handle_cross_chain_request(request).await?;
+                        Ok::<_, anyhow::Error>(response)
```

Comment:
Nit: `anyhow::Result::Ok(response)` avoids the turbofish :)

Reply:
Nice, good idea! (I didn't add this line, though; only moved it.)

Reply:
Surprisingly, that doesn't work, and it still says:
```
468 |                         anyhow::Result::Ok(response)
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `E` declared on the enum `Result`
```

Reply:
Oh I didn't realize they'd done that!  There's an optional type parameter for the error.  Yes you'll have to `anyhow::Result::<_>::Ok`.  No escaping the turbofish I suppose.